### PR TITLE
Adds V3 Market Contract and transactions

### PR DESCRIPTION
**Pull request by joshuahannan**
Adds TopShotMarketV3, which is similar to V2 in that it uses a capability to the user's moment collection, but it also imports the V1 market contract and stores a capability to the V1 sale collection so that a single call to the SaleCollection will check both the V1 SaleCollection and V3 SaleCollection

Adds tests for V3

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -363,7 +363,7 @@ and if they sent the correct amount, they get the Moment back.
 There are two versions of the Top Shot Market Contract.
```
Comment (sadief) [inline README.md:363]: Should we make reference to there being the intermediary `TopShotMarketV2` version - and why V3 was created on top of it?

File: contracts/TopShotMarketV3.cdc [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+                if let v1Market = self.marketV1Capability {
+                    let v1MarketRef = v1Market.borrow()!
+
+                    return <-v1MarketRef.purchase(tokenID: tokenID, buyTokens: <-buyTokens)
+                } else {
+                    panic("No token matching this ID for sale!")
+                }
+            }
+            
+            destroy buyTokens // THIS LINE NEEDS TO BE REMOVED ONCE THE BUG IN CADENCE IS FIXED
```
Comment (sadief) [inline contracts/TopShotMarketV3.cdc:235]: Just for context - what is the bug in Cadence?

File: transactions/marketV3/purchase_both_markets.cdc [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        } else if let marketV1CollectionRef = seller.getCapability(/public/topshotSaleCollection)
+            .borrow<&{Market.SalePublic}>() {
+            // purchase the moment
+            let purchasedToken <- marketV1CollectionRef.purchase(tokenID: momentID, buyTokens: <-self.purchaseTokens)
+
+            // deposit the purchased moment into the signer's collection
+            receiverRef.deposit(token: <-purchasedToken)
+
+        } else {
+            destroy self.purchaseTokens // REMOVE THIS
```
Comment (sadief) [inline transactions/marketV3/purchase_both_markets.cdc:51]: Is this related to the Cadence bug as well?

File: contracts/TopShotMarketV3.cdc (hunk 2 above)
Comment (joshuahannan) [reply contracts/TopShotMarketV3.cdc:235]: https://github.com/onflow/cadence/pull/1000
It'll be released in the next version of Cadence, and I'll update this after that is released

File: transactions/marketV3/purchase_both_markets.cdc (hunk 3 above)
Comment (joshuahannan) [reply transactions/marketV3/purchase_both_markets.cdc:51]: Yes, it is the same bug

File: lib/go/templates/marketV3_templates.go [hunk 4]
```diff
@@ -5,142 +5,144 @@ import (
...
-func GenerateCreateAndStartSaleV2Script(env Environment) []byte {
-	code := assets.MustAssetString(transactionsPath + createAndStartSaleV2Filename)
+func GenerateCreateAndStartSaleV3Script(env Environment) []byte {
+	code := assets.MustAssetString(transactionsPath + createAndStartSaleV3Filename)
 
 	return []byte(replaceAddresses(code, env))
 }
 
-// GenerateCancelSaleV2Script creates a cadence transaction that starts a sale by depositing
+// GenerateCancelSaleV3Script creates a cadence transaction that starts a sale by depositing
```
Comment (sadief) [inline lib/go/templates/marketV3_templates.go:56]: Just to double check - we would use this now over the `GenerateWithdrawFromSaleScript`  because we no longer need a a withdraw in order to delist a moment, right? (Do we need to update the comment for this one?)

File: README.md (hunk 1 above)
Comment (joshuahannan) [reply README.md:363]: yeah I can add a line

File: lib/go/templates/marketV3_templates.go (hunk 4 above)
Comment (joshuahannan) [reply lib/go/templates/marketV3_templates.go:56]: yep, I'll change the comment

File: contracts/TopShotMarketV3.cdc [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+                    .deposit(from: <-beneficiaryCut)
+                
+                // Deposit the remaining tokens into the owners vault
+                self.ownerCapability.borrow()!
+                    .deposit(from: <-buyTokens)
+
+                emit MomentPurchased(id: tokenID, price: price, seller: self.owner?.address)
+
+                // Return the purchased token
+                let boughtMoment <- self.ownerCollection.borrow()!.withdraw(withdrawID: tokenID) as! @TopShot.NFT
```
Comment (DavidRS91) [inline contracts/TopShotMarketV3.cdc:226]: Will there be any logic to cancel existing sales when a moment is transferred? Based on my novice understanding of cadence, I guess a sale would exist after transfer, and fail here when someone tries to make a purchase on a transferred moment?

File: lib/go/contracts/contracts.go [hunk 6]
```diff
@@ -65,14 +67,15 @@ func GenerateTopShotMarketContract(ftAddr, nftAddr, topshotAddr string) []byte {
...
+// GenerateTopShotMarketV3Contract returns a copy
+// of the third version TopShotMarketContract with the import addresses updated
+func GenerateTopShotMarketV3Contract(ftAddr, nftAddr, topshotAddr, marketAddr string) []byte {
 
-	marketCode := assets.MustAssetString(marketV2File)
+	marketCode := assets.MustAssetString(marketV3File)
 	codeWithNFTAddr := strings.ReplaceAll(marketCode, defaultNonFungibleTokenAddress, nftAddr)
 	codeWithTopshotAddr := strings.ReplaceAll(codeWithNFTAddr, defaultTopshotAddress, topshotAddr)
 	codeWithFTAddr := strings.ReplaceAll(codeWithTopshotAddr, defaultFungibleTokenAddress, ftAddr)
+	codeWithMarketV2Addr := strings.ReplaceAll(codeWithFTAddr, defaultMarketAddress, marketAddr)
```
Comment (DavidRS91) [inline lib/go/contracts/contracts.go:78]:
```suggestion
	codeWithMarketV3Addr := strings.ReplaceAll(codeWithFTAddr, defaultMarketAddress, marketAddr)
```

File: lib/go/contracts/contracts.go [hunk 7]
```diff
@@ -65,14 +67,15 @@ func GenerateTopShotMarketContract(ftAddr, nftAddr, topshotAddr string) []byte {
...
 
-	return []byte(codeWithFTAddr)
+	return []byte(codeWithMarketV2Addr)
```
Comment (DavidRS91) [inline lib/go/contracts/contracts.go:80]:
```suggestion
	return []byte(codeWithMarketV3Addr)
```

File: lib/go/templates/template_utils.go [hunk 8]
```diff
@@ -3,6 +3,8 @@ package templates
 import (
 	"fmt"
 	"strings"
+
+	_ "github.com/kevinburke/go-bindata"
```
Comment (DavidRS91) [inline lib/go/templates/template_utils.go:7]: Is this import needed?

File: lib/go/test/market_test.go [hunk 9]
```diff
@@ -871,24 +879,261 @@ func TestV2Market(t *testing.T) {
...
+		_ = tx.AddArgument(cadence.NewUInt64(3))
+		_ = tx.AddArgument(CadenceUFix64("50.0"))
+
+		signAndSubmit(
+			t, b, tx,
+			[]flow.Address{b.ServiceKey().Address, tokenAddr}, []crypto.Signer{b.ServiceKey().Signer(), tokenSigner},
+			false,
+		)
+
+		// Shoulf fail because the price is wrong
```
Comment (DavidRS91) [inline lib/go/test/market_test.go:1043]: Minor
```suggestion
		// Should fail because the price is wrong
```

File: transactions/marketV3/mint_and_purchase.cdc [hunk 10]
```diff
@@ -17,8 +18,8 @@ transaction(sellerAddress: Address, recipient: Address, tokenID: UInt64, purchas
         destroy minter
 
         let seller = getAccount(sellerAddress)
-        let topshotSaleCollection = seller.getCapability(/public/topshotSaleCollection)
-            .borrow<&{TopShotMarketV2.SalePublic}>()
+        let topshotSaleCollection = seller.getCapability(TopShotMarketV3.marketPublicPath)
+            .borrow<&{Market.SalePublic}>()
```
Comment (DavidRS91) [inline transactions/marketV3/mint_and_purchase.cdc:22]: Should this be v3?

File: contracts/TopShotMarketV3.cdc (hunk 5 above)
Comment (joshuahannan) [reply contracts/TopShotMarketV3.cdc:226]: There isn't logic for it in the smart contract, but we could update the transfer moment transaction to also cancel the sale if it exists

File: transactions/marketV3/mint_and_purchase.cdc (hunk 10 above)
Comment (joshuahannan) [reply transactions/marketV3/mint_and_purchase.cdc:22]: No, this is correct. The public interface for the two is exactly the same so we wanted to use it instead of creating a new one

File: lib/go/templates/template_utils.go (hunk 8 above)
Comment (joshuahannan) [reply lib/go/templates/template_utils.go:7]: No, but I was having a lot of difficulty with go modules and dependencies getting CI to pass properly, and this is the only way I was able to figure out how to make it work. 🤷

File: contracts/TopShotMarketV3.cdc (hunk 5 above)
Comment (joshuahannan) [reply contracts/TopShotMarketV3.cdc:226]: Do you think that you'd like to use a transaction like that?

File: transactions/marketV3/mint_and_purchase.cdc (hunk 10 above)
Comment (DavidRS91) [reply transactions/marketV3/mint_and_purchase.cdc:22]: Got it. Thanks for the explanation!

File: contracts/TopShotMarketV3.cdc (hunk 5 above)
Comment (DavidRS91) [reply contracts/TopShotMarketV3.cdc:226]: I think that could make sense - I'll defer to @sadief @EricLin2004 and @rrrkren here though.

**Review by DavidRS91 — APPROVED** [review]
LGTM!

File: contracts/TopShotMarketV3.cdc (hunk 5 above)
Comment (joshuahannan) [reply contracts/TopShotMarketV3.cdc:226]: We don't borrow a capability to the moment itself, so nothing would happen to a capability because it doesn't exist. If the moment that was for sale is transferred, then when someone tries to purchase it, the purchase would fail and would revert. If we just put the cancel sale in the transfer transaction, then we would avoid this in the top shot marketplace

File: contracts/TopShotMarketV3.cdc [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+    Description: Contract definitions for users to sell their moments
+
+    Marketplace is where users can create a sale collection that they
+    store in their account storage. In the sale collection, 
+    they can put their NFTs up for sale with a price and publish a 
+    reference so that others can see the sale.
+
+    If another user sees an NFT that they want to buy,
+    they can send fungible tokens that equal or exceed the buy price
+    to buy the NFT.  The NFT is transferred to them when
```
[the remark this answers is not in the document]
Comment (joshuahannan) [reply contracts/TopShotMarketV3.cdc:14]: That kind of introduces more complications. They'd also have to include a capability for their vault so the leftovers would be sent back to them. Also, as far as I know, we don't allow the users to change their prices without first canceling the sale, so I don't think we would ever encounter this

File: contracts/TopShotMarketV3.cdc [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+            } else if let v1Market = self.marketV1Capability {
+                let v1MarketRef = v1Market.borrow()!
+
+                return <-v1MarketRef.purchase(tokenID: tokenID, buyTokens: <-buyTokens)
+            }
+            
+            panic("No token matching this ID for sale!")
+        }
+
+        /// changePercentage changes the cut percentage of the tokens that are for sale
```
[the remark this answers is not in the document]
Comment (joshuahannan) [reply contracts/TopShotMarketV3.cdc:240]: Yeah, the cutPercentage is taken at purchase time, so if someone changes the percentage, it will apply to all the tokens in the sale. I thought we had decided that that was what we wanted. Is it not desired any more?

File: contracts/TopShotMarketV3.cdc (hunk 5 above)
Comment (EricLin2004) [reply contracts/TopShotMarketV3.cdc:226]: We should talk about this. Thanks for clarifying on specifics but the fact remains that we will have some listings in the marketplace that are no longer valid if we don't remove them, yes?

File: contracts/TopShotMarketV3.cdc (hunk 11 above)
[the remark this answers is not in the document]
Comment (EricLin2004) [reply contracts/TopShotMarketV3.cdc:14]: Yeah that's what I was pointing out, your comment here says:

>If another user sees an NFT that they want to buy,
>they can send fungible tokens that equal or exceed the buy price

Says equal or exceeds.

File: contracts/TopShotMarketV3.cdc (hunk 12 above)
[the remark this answers is not in the document]
Comment (EricLin2004) [reply contracts/TopShotMarketV3.cdc:240]: I don't remember why we set it this way though, do you?

It feels odd that you can change the cutPercentage of EXISTING listed items.

A malicious operator could wait for everyone to list and say the cutPercentage is 1% but after everything is listed they can set it to 50% or something?

[the remark this answers is not in the document]
Comment (rrrkren) [reply contracts/TopShotMarketV3.cdc:240]: maybe add a check in purchase to validate cut percentage?

[the remark this answers is not in the document]
Comment (rrrkren) [reply contracts/TopShotMarketV3.cdc:240]: what exactly does that achieve thou?
buyer: they pay the same amount regardless of cut percentage
seller: they would be the party that authorize the cut percentage change?

File: transactions/user/transfer_moment_v3_sale.cdc [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        // borrow a reference to the owner's collection
+        let collectionRef = acct.borrow<&TopShot.Collection>(from: /storage/MomentCollection)
+            ?? panic("Could not borrow a reference to the stored Moment collection")
+        
+        // withdraw the NFT
+        self.transferToken <- collectionRef.withdraw(withdrawID: withdrawID)
+
+        if let saleRef = acct.borrow<&TopShotMarketV3.SaleCollection>(from: TopShotMarketV3.marketStoragePath) {
+            saleRef.cancelSale(tokenID: withdrawID)
+        }
```
Comment (sadief) [inline transactions/user/transfer_moment_v3_sale.cdc:29]: Are we already cancelling the sale before a transaction here?

Comment (joshuahannan) [reply transactions/user/transfer_moment_v3_sale.cdc:29]: Yeah, this is transaction that cancels the sale at the same time as the transfer

File: contracts/TopShotMarketV3.cdc (hunk 5 above)
Comment (sadief) [reply contracts/TopShotMarketV3.cdc:226]: Can I clarify - The `cancelSale` already happens upon a transaction [here](https://github.com/dapperlabs/nba-smart-contracts/pull/118#discussion_r666547298) so confirming that there is no action required on this comment now? @EricLin2004 @joshuahannan

File: contracts/TopShotMarketV3.cdc (hunk 12 above)
[the remark this answers is not in the document]
Comment (sadief) [reply contracts/TopShotMarketV3.cdc:240]: @EricLin2004 @rrrkren @joshuahannan clarification: do we want to remove `changePercentage` entirely?

File: contracts/TopShotMarketV3.cdc (hunk 5 above)
Comment (joshuahannan) [reply contracts/TopShotMarketV3.cdc:226]: Yes, that is correct

File: contracts/TopShotMarketV3.cdc (hunk 12 above)
[the remark this answers is not in the document]
Comment (joshuahannan) [reply contracts/TopShotMarketV3.cdc:240]: Yeah, the seller is the one who sets the cut percentage, so it would only change it for their sales, which would only affect them. The buyer still pays the same price. I think we should keep the `changePercentage` function in case the top shot team decides to change it in the future

[the remark this answers is not in the document]
Comment (sadief) [reply contracts/TopShotMarketV3.cdc:240]: Just chatted with @EricLin2004 - he would like the `changePercentage` to be removed as we don't currently use it and likely will not (can't think of any reason that it would make sense for a seller to be able to change the cut percentage) and since the contracts are public we don't want to set a precedent for it that others may use as an example

[the remark this answers is not in the document]
Comment (sadief) [reply contracts/TopShotMarketV3.cdc]: Removed `changePercentage`

File: contracts/TopShotMarketV3.cdc [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+        /// Parameters: tokenID: the ID of the NFT to purchase
+        ///             buyTokens: the fungible tokens that are used to buy the NFT
+        ///
+        /// Returns: @TopShot.NFT: the purchased NFT
+        pub fun purchase(tokenID: UInt64, buyTokens: @FungibleToken.Vault): @TopShot.NFT {
+
+            // First check this sale collection for the NFT
+            if self.prices[tokenID] != nil {
+                assert(
+                    buyTokens.balance == self.prices[tokenID]!,
```
[the remark this answers is not in the document]
Comment (sadief) [reply contracts/TopShotMarketV3.cdc:202]: Resolved - removed comment

**Review by EricLin2004 — APPROVED** [review]

**Review by rrrkren — APPROVED** [review]
